main/acle.rst

```diff
@@ -2226,6 +2226,201 @@ or compiler-generated functions.
+   +--------------+-------------------------------------+---------------------+-------------------------------------+
+   | 350          | ``FEAT_SVE_SHA3``                   | sve_sha3            | ID_AA64ZFR0_EL1.SHA3 == 0b0001      |
+   +--------------+-------------------------------------+---------------------+-------------------------------------+
+   | 360          | ``FEAT_SME``                        | sme                 | ID_AA64PFR1_EL1.SME == 0b0001       |
+   +--------------+-------------------------------------+---------------------+-------------------------------------+
+   | 370          | ``FEAT_MTE``                        | mte                 | ID_AA64PFR1_EL1.MTE >= 0b0001       |
+   +--------------+-------------------------------------+---------------------+-------------------------------------+
+   | 380          | ``FEAT_MTE2``                       | mte2                | ID_AA64PFR1_EL1.MTE >= 0b0010       |
+   +--------------+-------------------------------------+---------------------+-------------------------------------+
+   | 390          | ``FEAT_MTE3``                       | mte3                | ID_AA64PFR1_EL1.MTE >= 0b0010       |
+   +--------------+-------------------------------------+---------------------+-------------------------------------+
+   | 400          | ``FEAT_SB``                         | sb                  | ID_AA64ISAR1_EL1.SB == 0b0001       |
+   +--------------+-------------------------------------+---------------------+-------------------------------------+
+   | 410          | ``FEAT_SSBS``                       | ssbs                | ID_AA64PFR1_EL1.SSBS == 0b0001      |
+   +--------------+-------------------------------------+---------------------+-------------------------------------+
+   | 420          | ``FEAT_SSBS2``                      | ssbs2               | ID_AA64PFR1_EL1.SSBS == 0b0010      |
+   +--------------+-------------------------------------+---------------------+-------------------------------------+
+   | 430          | ``FEAT_BTI``                        | bti                 | ID_AA64PFR1_EL1.bt == 0b0010        |
+   +--------------+-------------------------------------+---------------------+-------------------------------------+
+   | max          |                                     | cortex-A55          | MIDR_EL1.Implementer == 0x41 &&     |
+   |              |                                     |                     | MIDR_EL1.PartNum == 0xD05           |
+   +--------------+-------------------------------------+---------------------+-------------------------------------+
+
+Selection
+----------
+
+The following rules shall be followed by all implementations.
+Implementation of the selection algoritm is platform dependent.
+
+Only those versions could be considered where all dependent features are available.
+
+Rules of version selection are in order:
+
+#. If possible select the CPU implementation specific version else
```

Comment:
Agree, in case of bigLITTLE the cpu implementation selection is complicated so I add a note about it. I don't think is will be widely used but I would keep it in the spec. On a given platforms it might be disabled in compile time.

ifunc resolver could be used I think but the resolver might use the  cpu feature register emulation.
https://www.kernel.org/doc/html/latest/arm64/cpu-feature-registers.html

> i think the specification does not make it clear that for the lifetime of a given process only one variant is selected.
Good catch, I make it clear.

Reply:
cpu feature register emulation does not solve all problems: ifuncs are problematic if there are any selection rules that may change value at runtime. in this case MIDR is different across cores so if that is used for selection then you have to ensure it is only read once and that result is used consistently throughout the process, which i dont know how to do with ifuncs. (reading MIDR multiple times can give different results if the process gets scheduled to different cores).

Reply:
yeah, non-homogeneous SoCs are a PITA. for Android's libc with arm64 the only variant we have right now is MTE, so we just use hwcap there and assume that no-one would ship a chip where only some cores support it: https://android.googlesource.com/platform/bionic/+/refs/heads/master/libc/arch-arm64/dynamic_function_dispatch.cpp

but that might be more problematic for less fundamental features.

/me wonders whether we should change the CDD to outlaw SoCs where hwcap doesn't report the same value for every core?

Reply:
Kernel ensures the provided HWCAPs\CPU Feature Registers are consistent across the cores and supported by the current kernel.
The only exception is MIDR in the CPU Feature Registers.
Seems out there only a few cases that really cares about the actual CPU so maybe dropping this option won't cause any trouble later.